pkgs/development/python-modules/pyswip/default.nix

```diff
@@ -0,0 +1,33 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+, pytestcov
+, pytest
+, swiProlog
+}:
+
+buildPythonPackage rec {
+  pname = "pyswip";
+  version = "0.2.9";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "4591414006340ad8f033f9f0ee4c4c9151cd5a309ed19741a9c79b9ffce8b58e";
+  };
+
+  propagatedNativeBuildInputs = [ swiProlog ];
+  checkInputs = [ pytest pytestcov ];
+  postFixup = ''
+  set -o pipefail -o errexit
+  path="$(python -c 'import pyswip.core; print(pyswip.core._path)')"
+  home_dir="$(python -c 'import pyswip.core; print(pyswip.core.SWI_HOME_DIR)')"
+  find $out -name core.py | xargs -n 1 sed -i "s:(_path, SWI_HOME_DIR) = _findSwipl():(_path, SWI_HOME_DIR) = ('$path', '$home_dir'):"
```

Comment:
do you mind do this as part of the patchPhase? like:

```
postPatch = ''
  substituteInplace pyswip/core.py \
    --replace "(_path, SWI_HOME_DIR) = _findSwipl()" "_path, SWI_HOME_DIR) = ('${swiProlog}/bin', '${swiProlog}')"
'';
```
I'm assuming this interacts with the swiprolog package

Reply:
Alas, it's not quite so simple. When built in Nix (and I assume on most systems), _findSwipl outputs ("/nix/store/hash-swi-prolog-8.1.4/lib/swipl/lib/x86_64-linux/libswipl", "/nix/store/hash-swi-prolog-8.1.4/lib/swipl"). I'm not sure how to get these paths without actually running the (quite complex) Python code making up _findSwipl().

Reply:
looking through their login, _path -> libswipl.so and SWI_HOME_DIR -> PLBASE key of `swipl --dump-runtime-variables". The resulting patch to make this work would be:

```
{ lib
, buildPythonPackage
, fetchPypi
, hostPlatform
, pytestcov
, pytest
, swiProlog
}:

buildPythonPackage rec {
  pname = "pyswip";
  version = "0.2.9";

  src = fetchPypi {
    inherit pname version;
    sha256 = "4591414006340ad8f033f9f0ee4c4c9151cd5a309ed19741a9c79b9ffce8b58e";
  };

  checkInputs = [ pytest pytestcov ];
  postPatch = ''
    substituteInPlace pyswip/core.py \
      --replace "(_path, SWI_HOME_DIR) = _findSwipl()" "(_path, SWI_HOME_DIR) = ('${swiProlog}/lib/libswipl.so', '${swiProlog}/lib/swipl/')"
  '';

  meta = with lib; {
    description = "PySwip enables querying SWI-Prolog in your Python programs";
    homepage = https://github.com/yuce/pyswip;
    license = licenses.mit;
    maintainers = [ maintainers.catern ];
  };
}
```

Reply:
```
[nix-shell:~/.cache/nix-review/pr-72803]$ nix-shell -p "with import ./nixpkgs {}; python37Packages.pyswip"
$ python
>>> from pyswip import Functor, Variable, Query, call
>>>
>>> assertz = Functor("assertz", 1)
>>> father = Functor("father", 2)
>>> call(assertz(father("michael","john")))
1
>>> call(assertz(father("michael","gina")))
1
>>> X = Variable()
>>>
>>> q = Query(father("michael",X))
>>> while q.nextSolution():
...      print("Hello,", X.value)
...
Hello, john
Hello, gina
>>> q.closeQuery()
```